### PR TITLE
feat(swarm): allow configuration to idle connection timeout

### LINKED ISSUE #4121: swarm: Introduce an `idle_connection_timeout` config option
## Description

Original discussion thread: https://github.com/libp2p/rust-libp2p/issues/3844

We want an `idle_connection_timeout` config on the `SwarmBuilder` that triggers once a connection reports `KeepAlive::No`. The current logic is here: https://github.com/libp2p/rust-libp2p/blob/524bfc2130ec06a4f2d586a8716593df5419e6d2/swarm/src/connection.rs#L339-L356

I think what we can do here is utilize the already existing `Shutdown::Later` variant and simply set it to the configured timeout in the `KeepAlive::No` branch.

By default, the timeout should be 0 seconds.

## Motivation

Several usecases like tests, examples, benchmarks and more client-server style applications often want to keep connections alive even if there currently isn't any work happening on the connection.

This will unblock https://github.com/libp2p/rust-libp2p/pull/4029. At the same time with introducing this feature, we should deprecate `keep_alive::Behaviour` and recommend users to simply set the new configuration option to:

- `u64::MAX` if they want to keep the connection alive "forever"
- a value that makes sense in their context

For tests for example, we would probably be fine to set the value to something like 10 seconds. We can do that here: https://github.com/libp2p/rust-libp2p/blob/master/swarm-test/src/lib.rs#L221

As a result, we can likely remove the usage of `keep_alive::Behaviour` from several tests.

## Are you planning to do it yourself in a pull request?

Maybe but would appreciate help.

### PR DESCRIPTION
**Pull request by startup-dreamer**
## Description

Previously, a connection would be shut down immediately as soon as its `ConnectionHandler` reports `KeepAlive::No`. As we have gained experience with libp2p, it turned out that this isn't ideal.

For one, tests often need to keep connections alive longer than the configured protocols require. Plus, some usecases require connections to be kept alive in general.

Both of these needs are currently served by the `keep_alive::Behaviour`. That one does essentially nothing other than statically returning `KeepAlive::Yes` from its `ConnectionHandler`.

It makes much more sense to deprecate `keep_alive::Behaviour` and instead allow users to globally configure an `idle_conncetion_timeout` on the `Swarm`. This timeout comes into effect once a `ConnectionHandler` reports `KeepAlive::No`. To start with, this timeout is 0. Together with https://github.com/libp2p/rust-libp2p/issues/3844, this will allow us to move towards a much more aggressive closing of idle connections, together with a more ergonomic way of opting out of this behaviour.

Fixes #4121.

## Notes & open questions

~am I in the right direction also we might add tests for this feature.~

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: Two things:

- I'd prefer to not use Option but assign a default value.
- Replace u64 with Duration to be expressive and delete the comment.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: This looks correct but it would be great to have a test for it!

**Review by thomaseizinger — COMMENTED** [review]
Thanks! Looks like a good start!

I'd like us to use this directly in `libp2p-swarm-test` and see which tests we can remove `keep_alive::Behaviour` from, should be all of them but maybe something else turns up!

File: swarm/src/connection.rs
Comment (thomaseizinger) [reply swarm/src/connection.rs]: We can also remove the `connection` from this name as this entire struct is about a connection.

**Comment by startup-dreamer** [comment]
Hy @thomaseizinger can you guide me through how I am gonna remove `keep_alive::Behaviour` tests can't wrap my head around this.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: I think there is a constant for 0.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: This test should cover that after 5 seconds, we shut the connection down.

You can use `dummy::ConnectionHandler` which never does anything.

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs]: We need to make this configurable from the `SwarmBuilder`.

**Comment by thomaseizinger** [comment]
> Hy @thomaseizinger can you guide me through how I am gonna remove `keep_alive::Behaviour` tests can't wrap my head around this.

Sure!

So the context is that with the current code, a connection will be immediately shut down as soon as a `ConnectionHandler` reports `KeepAlive::No`.

In many tests, we only use a single `NetworkBehaviour` which is quite an artificial scenario and as such, we often run into the problem that connections get shut down too quickly despite us still wanting to use them in the test.

With this new feature, we can delay that shutdown which in turn allows us to remove `keep_alive::Behaviour` from many tests.

I'd take the following approach:

- Put a `#[deprecated]` attribute on `keep_alive::Behaviour`
- Run cargo check / cargo clippy with `--all-targets`
- Go through each usage that is a test and remove it. Most of the time, this will be as simple as deleting a field from a struct
- Run the test. Likely, it will fail.
- Check where the `Swarm` in this particular test gets constructed.
   - For the usages coming from `Swarm::new_ephemeral`, you can extend the `SwarmBuilder` in there to configure a 10 sec keep alive timeout. That should be more than plenty for each test. This will likely fix a lot of tests because we extensively use that test harness in our codebase.
  - For any other usage, simply extend that `SwarmBuilder` with a 10 second timeout.

Does that make sense? You can try yourself on one or two tests and then push a commit and we can look at it together :)

File: swarm/src/connection.rs
Comment (mxinden) [reply swarm/src/connection.rs]: https://doc.rust-lang.org/std/time/struct.Duration.html#associatedconstant.ZERO

**Review by mxinden — COMMENTED** [review]
Thank you for the help here.

**Comment by startup-dreamer** [comment]
> Thank you for the help here.

I still wan't to work on this however I am busy in some other work so if anyone wanna take this he can.
If not I will complete this as soon as possible.

**Comment by thomaseizinger** [comment]
> > Thank you for the help here.
> 
> I still wan't to work on this however I am busy in some other work so if anyone wanna take this he can.
> If not I will complete this as soon as possible.

Thank you! No pressure from our end, your help is much appreciated :)

**Comment by thomaseizinger** [comment]
@startup-dreamer I've pushed some commits that should help move this forward. In particular, I've added the deprecation attribute and converted the first test. Most of the others should be similar :)

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs:1531]: We need to mention this addition in the changelog.

**Comment by startup-dreamer** [comment]
> @startup-dreamer I've pushed some commits that should help move this forward. In particular, I've added the deprecation attribute and converted the first test. Most of the others should be similar :)

graet help sir will try to move forward from here.

**Comment by startup-dreamer** [comment]
@thomaseizinger should i wait for you to review each refactor I does or do a bunch of refactoring and then take the review

**Review by thomaseizinger — COMMENTED** [review]
Looks good so far :)

**Comment by thomaseizinger** [comment]
> @thomaseizinger should i wait for you to review each refactor I does or do a bunch of refactoring and then take the review

Looks good so far, you are on the right track :)

File: examples/metrics/src/main.rs
Comment (thomaseizinger) [inline examples/metrics/src/main.rs]: See this comment!

In this example, we want to retain the `identify` and the `ping` behaviour, otherwise the metrics don't show anything interesting :)

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs]: We will have to keep this otherwise it is a breaking change.

**Comment by thomaseizinger** [comment]
I'd prefer if you would avoid force-pushing. See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#we-squash-merge-pull-requests.

**Comment by startup-dreamer** [comment]
> I'd prefer if you would avoid force-pushing. See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#we-squash-merge-pull-requests.

My bad, I committed the file in which I was working. I should have squash-merged it instead. I will keep this in mind.

**Comment by thomaseizinger** [comment]
> My bad, I committed the file in which I was working.

As long as you haven't pushed it, you are free to edit commits locally. But once you've pushed them, just make new commits on top :)

It is easier to review because Github is then able to show a nice diff since the last review!

**Comment by startup-dreamer** [comment]
@thomaseizinger This should do it, Only two things remains
- The changelog entry
- resolving the conflict's

will do them once you reviews these changes.

File: examples/metrics/src/main.rs
Comment (thomaseizinger) [inline examples/metrics/src/main.rs]: That isn't quite true :)

The `dummy::Behaviour` does nothing so you can remove it from this struct. This comment should move to the `idle_connection_timeout()` function call and we should set it to a much higher value, like 1 minute or something. See, what `keep_alive::Behaviour` used to do here is keep all connections alive indefinitely. That would allow `ping::Behaviour` to keep pinging and this example would showcase the metrics collected as a result. Now we are replacing `keep_alive::Behaviour` with a different mechanism :)

Let me know if this makes sense!

File: examples/ping-example/src/main.rs
Comment (thomaseizinger) [inline examples/ping-example/src/main.rs]: Let's set this to one minute here as well and leave a comment:

 ```suggestion
            .idle_connection_timeout(Duration::from_secs(60)) // For illustrative purposes, keep idle connections alive for a minute so we can observe a few pings.
```

File: examples/rendezvous/src/bin/rzv-discover.rs
Comment (thomaseizinger) [inline examples/rendezvous/src/bin/rzv-discover.rs]: No need to include this if we have other behaviours! :)

File: examples/rendezvous/src/bin/rzv-identify.rs
Comment (thomaseizinger) [inline examples/rendezvous/src/bin/rzv-identify.rs]: Same here, no need to include this!

File: misc/allow-block-list/src/lib.rs
Comment (thomaseizinger) [inline misc/allow-block-list/src/lib.rs]: This can also be simplified by removing this struct altogether and directly referencing `super::Behaviour<S>` in the tests. You might need to configure an idle connection timeout for the tests to keep passing!

File: misc/connection-limits/src/lib.rs
Comment (thomaseizinger) [inline misc/connection-limits/src/lib.rs]: Same here as for the `allow-block-list` module.

File: protocols/identify/tests/smoke.rs
Comment (thomaseizinger) [inline protocols/identify/tests/smoke.rs]: Delete the `Behaviour` struct and reference the `identify::Behaviour` directly in the test.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]:
```suggestion
                (_, KeepAlive::No) if idle_timeout == Duration::ZERO => {
                	*shutdown = Shutdown::Asap;
              	}
              	(_, KeepAlive::No) => {
                	let deadline = Instant::now() + *idle_timeout;
                        *shutdown = Shutdown::Later(Delay::new(*idle_timeout), deadline);
              	}
```

Might be more elegant to handle this in two branches because it results in less indentation.

File: swarm/src/connection.rs
Comment (thomaseizinger) [inline swarm/src/connection.rs]: To actually test this feature fully, this test should wait for the configured idle-timeout, poll again and we should receive a keep-alive timeout error.

I'd suggest to change it to a `#[tokio::test]` for that and use `future::poll_fn` instead of `poll_noop_waker` to make sure we properly register the waker.

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs]:
```suggestion
    note = "Configure an appropriate idle connection timeout via `SwarmBuilder::idle_connection_timeout` instead. To keep connections alive 'forever', use `Duration::from_secs(u64::MAX)`."
```

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs]: With this feature, I think `new_test_swarm` is always called with `dummy::ConnectionHandler`! This means we can simplify these tests by removing the parameter and constructing `dummy::ConnectionHandler` inside.

Some of the comments around here will have to be updated aswell!

File: swarm/src/lib.rs
Comment (thomaseizinger) [inline swarm/src/lib.rs]: If we end up configuring this in every test, we might as well do it inside `new_test_swarm`!

**Review by thomaseizinger — COMMENTED** [review]
Great progress!

I've made some comments on where I think we can improve a few more things, let me know if those make sense :)

Looking forward to landing this! It is an important piece within the greater effort of improving handling of idle connections. You can read more about it here if you are interested: https://github.com/libp2p/rust-libp2p/discussions/3353#discussioncomment-6144804

File: swarm/src/connection.rs
Comment (mxinden) [inline swarm/src/connection.rs]: Let's not use the zero value (i.e. `Duration::ZERO`) as a signal for absence. Rust has sum-types, e.g. `Option`. Let's instead have `duration` be an `Option<Duration>` and check if `duration` is `None` here.

**Review by mxinden — COMMENTED** [review]
Small drive-by comment.

Direction looks good to me.

Thank you for the work. Still many comments from Thomas open. Let me know once you need another round of review.

File: swarm/src/connection.rs
Comment (thomaseizinger) [reply swarm/src/connection.rs]: My thinking would have been the opposite. You can't avoid the case that the `Duration` is 0 in the type-system, even with an `Option`, so we might as well use that but optimise it to not unnecessarily create a `Delay`. Thoughts?

Comment (thomaseizinger) [reply swarm/src/connection.rs]: Looking at the code for `Delay`, creating one with `0` is not free so I think it is worth special-casing this.

**Comment by thomaseizinger** [comment]
Friendly ping @startup-dreamer. Are you still working on this?

**Comment by thomaseizinger** [comment]
@mxinden I spent some cycles on this to push it over the finish line as there wasn't much missing. Please review!

File: swarm/src/connection.rs
Comment (mxinden) [inline swarm/src/connection.rs:384]: If I understand the match arm further above correctly (copied below for ease of read) the following situation can happen. The user sets `idle_connection_timeout` to a large value, say 1h. One `ConnectionHandler` returns `KeepAlive::Until` with a short timeout, say 1s. In case nothing else happens on the connection, it will be closed after 1s.

One could argue that this is the correct behavior. I don't think this is intuitive. I can not think of a alternative, i.e. a more intuitive mechanism.

``` rust
                (Shutdown::Later(timer, deadline), KeepAlive::Until(t)) => {
                    if *deadline != t {
                        *deadline = t;
                        if let Some(dur) = deadline.checked_duration_since(Instant::now()) {
                            timer.reset(dur)
                        }
                    }
                }
```

Should we deprecate `KeepAlive::Until` with this pull request already, to prevent the above footgun to exit for long?

**Review by mxinden — COMMENTED** [review]
One comment, otherwise this looks good to me.

Let's have this run on `am6-rust` (IPFS bootstrap node staging server) once merged into master before cutting a release. Will be deployed automatically via our nightly job once merged.

File: swarm/src/connection.rs
Comment (thomaseizinger) [reply swarm/src/connection.rs:384]: I think deprecating `Until` takes a bit more work, at least if we want to remove all usages in our codebase at the same time.

Should we have the 1h start ticking once the `Until` finishes?

Comment (mxinden) [reply swarm/src/connection.rs:384]: > Should we have the 1h start ticking once the `Until` finishes?

Instead of keeping the connection alive for the duration of the sum of the two (i.e. `KeepAlive::Until` + `idle_connection_timeout`) , how about for the duration of the maximum of the two.

From both perspectives (i.e. `KeepAlive::Until` and `idle_connection_timeout`), the connection is kept alive at least as much as configured, potentially longer.

I would find this more intuitive and I assume this is easier to implement, as it would pretty much only require changing a `!=` for a `<`.

@thomaseizinger thoughts?

Comment (thomaseizinger) [reply swarm/src/connection.rs:384]: Yeah that sounds good to me! :)

I'll fix that later so hopefully we can land this soon 😍

**Comment by thomaseizinger** [comment]
> Let's have this run on `am6-rust` (IPFS bootstrap node staging server) once merged into master before cutting a release. Will be deployed automatically via our nightly job once merged.

With or without setting a custom timeout?

File: swarm/Cargo.toml
Comment (thomaseizinger) [inline swarm/Cargo.toml]:
```suggestion
libp2p-kad = { path = "../protocols/kad" } # Using `path` here because this is a cyclic dev-dependency which otherwise breaks releasing.
```

**Comment by mxinden** [comment]
> > Let's have this run on `am6-rust` (IPFS bootstrap node staging server) once merged into master before cutting a release. Will be deployed automatically via our nightly job once merged.
> 
> With or without setting a custom timeout?

I suggest without, to validate that the status-quo doesn't change.

**Review by mxinden — APPROVED** [review]
Cool. Thank you for the tests!

**Comment by mxinden** [comment]
@thomaseizinger can you address the CI failures? Otherwise this is ready to merge from my side.

**Comment by thomaseizinger** [comment]
> @thomaseizinger can you address the CI failures? Otherwise this is ready to merge from my side.

Done! I'll send it then :)

**Pull request merged**
